plugins/gitlab/server/GitLabIssueProvider.ts

```diff
@@ -0,0 +1,188 @@
+import { IssueSource } from "@shared/schema";
+import { IntegrationService, IntegrationType } from "@shared/types";
+import Logger from "@server/logging/Logger";
+import { Integration, IntegrationAuthentication } from "@server/models";
+import { BaseIssueProvider } from "@server/utils/BaseIssueProvider";
+import { GitLab } from "./gitlab";
+import { sequelize } from "@server/storage/database";
+
+export class GitLabIssueProvider extends BaseIssueProvider {
+  constructor() {
+    super(IntegrationService.GitLab);
+  }
+
+  async fetchSources(
+    integration: Integration<IntegrationType.Embed>
+  ): Promise<IssueSource[]> {
+    await integration.reload({
+      include: [
+        {
+          model: IntegrationAuthentication,
+          as: "authentication",
+          required: true,
+        },
+      ],
+    });
+
+    if (!integration.authentication) {
+      Logger.warn("GitLab integration without authentication");
+      return [];
+    }
+
+    const sources: IssueSource[] = [];
+
+    try {
+      const projects = await GitLab.getProjects(
+        integration.authentication.token
+      );
+
+      sources.push(
+        ...projects.map<IssueSource>((project) => ({
+          id: String(project.id),
+          name: project.name,
+          owner: {
+            id: String(project.namespace.id),
+            name: project.namespace.full_path,
+          },
+          service: IntegrationService.GitLab,
+        }))
+      );
+    } catch (err) {
+      Logger.warn("Failed to fetch projects from GitLab", err);
+    }
+
+    return sources;
+  }
+
+  async handleWebhook({
+    payload,
+    headers,
+  }: {
+    payload: Record<string, unknown>;
+    headers: Record<string, unknown>;
+  }) {
+    const hookId = headers["X-Gitlab-Webhook-UUID"] as string;
+    const eventName = payload.event_name as string;
+
+    if (!eventName) {
+      Logger.warn(
+        `Received GitLab webhook without event name; hookId: ${hookId}, eventName: ${eventName}`
+      );
+      return;
+    }
+
+    switch (eventName) {
+      case "project_update":
+      case "project_transfer":
+      case "project_rename":
+        await this.updateProject(payload);
+        break;
+      case "project_create":
+      case "project_destroy":
+        await this.createOrDestroyProject(payload, eventName);
+        break;
+
+      default:
+        break;
+    }
+  }
+
+  private async createOrDestroyProject(
+    payload: Record<string, any>,
+    eventName: string
+  ) {
+    let integration: Integration<IntegrationType.Embed> | null = null;
+
+    for (const owner of payload.owners) {
+      integration = (await Integration.findOne({
+        where: {
+          service: IntegrationService.GitLab,
+          "settings.github.installation.account.name": owner.name,
```

Comment:
@tommoor, this is an admittedly a flawed way to retrieve the integration, since this value is just a display name and users can (and do) change it, which would break the lookup.

An alternative would be to update all instances of the project across all GitLab integrations, though that approach has its own drawbacks. Gitlab doesn't really provide much identifying information for the project owners in these hooks.

thoughts?